File: dynamic_programming/linear_dp/no_chain/0650_2_keys_keyboard.js
/*
There is only one character 'A' on the screen of a notepad. You can perform 
one of two operations on this notepad for each step:

1. Copy All: You can copy all the characters present on the screen.
2. Paste: You can paste the characters which are copied last time.

Given an integer n, return the minimum number of operations to get the 
character 'A' exactly n times on the screen.

1 <= n <= 1000
*/

// https://leetcode.com/problems/2-keys-keyboard/description/

// if n === 1, return 0
// for any n > 2, the first and second steps must be copy and paste
// the least efficient way is to copy and paste n-1 times: minSteps(n) <= n
// (*2) -> c+p; (*3) -> c+p+p; (*4 = *2 *2) -> c+p+c+p; (*5) -> c+p+p+p+p...
// if n is a prime, the only way to get n is to copy and paste n-1 times
// let dp[i] be the minSteos(i)
// for n = k1 * k2 * ... * ki, dp(n) = dp(k1) + dp(k2) + ... + dp(ki)
var minSteps = function (n) {
  let dp = new Array(n + 1).fill(0);
  // corner case
  dp[1] = 0;
  dp[2] = 2;
  dp[3] = 3;
  // fill the dp array
  for (let i = 2; i <= n; i++) {
    // initialize dp[i] to i (the maximum steps to get i)
    dp[i] = i;
    for (let k = 2; k * k <= i; k++) {
      // if i is not a prime, dp[i] = dp[k] + dp[i/k]
      if (i % k === 0) {
        dp[i] = dp[k] + dp[i / k];
        break;
      }
    }
  }
  return dp[n];
};

// memoization
var minSteps = function (n) {
  let memo = new Map();
  var dp = function (n) {
    // base cases
    if (n === 1) return 0;
    if (n === 2) return 2;
    if (n === 3) return 3;
    if (memo.has(n)) return memo.get(n);
    // recursive cases
    for (let i = 2; i * i <= n; i++) {
      if (n % i === 0) {
        memo.set(n, dp(i) + dp(n / i));
        return memo.get(n);
      }
    }
    memo.set(n, n);
    return memo.get(n);
  };
  return dp(n);
};

File: dynamic_programming/linear_dp/no_chain/0650_2_keys_keyboard.py
# There is only one character 'A' on the screen of a notepad. You can perform 
# one of two operations on this notepad for each step:

# 1. Copy All: You can copy all the characters present on the screen.
# 2. Paste: You can paste the characters which are copied last time.

# Given an integer n, return the minimum number of operations to get the 
# character 'A' exactly n times on the screen.

# 1 <= n <= 1000

# https://leetcode.com/problems/2-keys-keyboard/

# math
# when n = 1, ans = 0 (no operation needed)
# when n > 1, the maximum ops is n (copy 1 time and paste n - 1 times)
# to lower ops: copy more than 1 time

# 如果我们将「1 次 Copy All + x - 1 次 Paste」看做一次“动作”的话,那么 一次“动作”所产生的效果就是将原来的字符串变为原来的 x 倍。

# 最终的最小操作次数方案可以等价以下操作流程：
# 起始对长度为 1 的记事本字符进行 1 次 Copy All + k1−1 次 Paste 操作（消耗次数为 k1，得到长度为 k1 的记事本长度）；
# 对长度为为 k1的记事本字符进行 1 次 Copy All + k2−1 次 Paste 操作（消耗次数为 k1 + k2，得到长度为 k1 * k2的记事本长度）
# ...
# 最终经过 x 次“动作”之后，得到长度为 n 的记事本长度，即有：
# n = k1 * k2 * ... * kx 
# 问题转化为：如何对 n 进行拆分，可以使得 k = k1 + k2 + ... + kx 最小。
# 当所有 ki 都是 n 的质因数时，k 达到最小 （for a, b >= 2: (a - 1)(b - 1) >= 1 -> ab >= a + b）
class Solution:
    def minSteps(self, n: int) -> int:
        ans = 0
        for i in range(2, isqrt(n) + 1):
            while n % i == 0:
                ans += i
                n //= i
        return ans if n == 1 else ans + n
